stdlib/public/core/EmbeddedRuntime.swift

```diff
@@ -233,11 +241,18 @@ public func swift_endAccess(buffer: UnsafeMutableRawPointer) {
 
 @_silgen_name("swift_once")
 public func swift_once(predicate: UnsafeMutablePointer<Int>, fn: (@convention(c) (UnsafeMutableRawPointer)->()), context: UnsafeMutableRawPointer) {
-  // TODO/FIXME: The following only works in single-threaded environments.
-  if predicate.pointee == 0 {
-    predicate.pointee = 1
+  if loadAcquire(predicate) < 0 { return }
```

Comment:
If the only valid values are -1, 0, and 1, then it might be good to raise a fatal error for other values, at least in debug builds.

Reply:
Added!

Reply:
I'll be interested to see how the AVR back end handles this! I'm sort of assuming atomics are a noop in AVR! (all single core machines)

Reply:
> AVR
I assume LLVM will lower all these into library calls, something like `__atomic_load`, `__atomic_store`, `__atomic_compare_exchange`, and expect that you either link compiler-rt that implements them, or that you provide your own definitions of them.

Though single-core doesn't necessarily mean single-threaded, but I assume all sane AVR setups are actually single-threaded?

Reply:
Yeah. Someone might contradict me but I'd say the only possible preemption on an AVR CPU is interrupt handlers.

I think we'll see how it lowers if and when we link to the new embedded runtime. And if the AVR backend doesn't have atomics it'll be an obvious compiler crash and an easy patch.